File: running-app-backend/social/models/post_comment.py
import uuid
from django.db import models
from django.core.validators import MaxLengthValidator
from django.db.models.signals import post_save, post_delete
from django.dispatch import receiver

class PostComment(models.Model):
    id = models.UUIDField(primary_key=True, default=uuid.uuid4, db_index=True)
    user = models.ForeignKey("account.Activity", on_delete=models.CASCADE)
    content = models.TextField()
    created_at = models.DateTimeField(auto_now_add=True)
    
    class Meta:
        abstract = True
        ordering = ["-created_at"]

    def __str__(self):
        return f"{self.user} - {self.created_at}"

class ClubPostComment(PostComment):
    post = models.ForeignKey(
        "social.ClubPost", on_delete=models.CASCADE, related_name="comments")
    
    def __str__(self):
        return f"{self.user.user.name} - {self.created_at} - Post: {self.post.id}"

class EventPostComment(PostComment):
    post = models.ForeignKey(
        "social.EventPost", on_delete=models.CASCADE, related_name="comments")

    def __str__(self):
        return f"{self.user.user.name} - {self.created_at} - Post: {self.post.id}"

class ActivityRecordPostComment(PostComment):
    # activity_record = models.ForeignKey(
    #     "activity.ActivityRecord", on_delete=models.CASCADE, related_name="comments")
    post = models.ForeignKey(
        "activity.ActivityRecord", on_delete=models.CASCADE, related_name="comments")
        
    def __str__(self):
        return f"{self.user.user.name} - {self.created_at} - Activity Record: {self.post.id}"

@receiver(post_save, sender=ClubPostComment)
def update_club_post_comments(sender, instance, created, **kwargs):
    if created:
        post = instance.post
        post.total_comments += 1
        post.save()

@receiver(post_save, sender=EventPostComment)
def update_event_post_comments(sender, instance, created, **kwargs):
    if created:
        post = instance.post
        post.total_comments += 1
        post.save()

@receiver(post_save, sender=ActivityRecordPostComment)
def update_activity_record_post_comments(sender, instance, created, **kwargs):
    if created:
        post = instance.post
        post.total_comments += 1
        post.save()

@receiver(post_delete, sender=ClubPostComment)
def delete_club_post_comments(sender, instance, **kwargs):
    post = instance.post
    post.total_comments -= 1
    post.save()

@receiver(post_delete, sender=EventPostComment)
def delete_event_post_comments(sender, instance, **kwargs):
    post = instance.post
    post.total_comments -= 1
    post.save()

@receiver(post_delete, sender=ActivityRecordPostComment)
def delete_activity_record_post_comments(sender, instance, **kwargs):
    post = instance.post
    post.total_comments -= 1
    post.save()

File: running_app/lib/models/social/post_comment.dart
import 'package:running_app/models/account/activity.dart';
import 'package:running_app/models/account/user_abbr.dart';
import 'package:running_app/models/activity/activity_record.dart';
import 'package:running_app/models/social/post.dart';

class PostComment {
  final String? id;
  final UserAbbr? user;
  final String? content;
  final String? createdAt;

  PostComment({
    required this.id,
    required this.user,
    required this.content,
    required this.createdAt,
  });

  PostComment.fromJson(Map<String, dynamic> json)
      : id = json['id'],
        user = UserAbbr.fromJson(json['user']),
        content = json['content'],
        createdAt = json['created_at'];

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user': user?.id,
      'content': content,
    };
  }

  @override
  String toString() {
    return '$user - $createdAt';
  }
}

class ClubPostComment extends PostComment {
  final String? post;

  ClubPostComment({
    String? id,
    UserAbbr? user,
    String? content,
    String? createdAt,
    required this.post,
  }) : super(
      id: id,
      user: user,
      content: content,
      createdAt: createdAt);

  ClubPostComment.fromJson(Map<String, dynamic> json)
      : post = json['post'],
        super.fromJson(json);

  @override
  Map<String, dynamic> toJson() {
    Map<String, dynamic> superJson = super.toJson();
    superJson['post_id'] = post;
    return superJson;
  }

  @override
  String toString() {
    return '${user} - $createdAt - Post: ${post}';
  }
}

class EventPostComment extends PostComment {
  final String? post;

  EventPostComment({
    String? id,
    UserAbbr? user,
    String? content,
    String? createdAt,
    required this.post,
  }) : super(
      id: id,
      user: user,
      content: content,
      createdAt: createdAt
  );

  EventPostComment.fromJson(Map<String, dynamic> json)
      : post = json['post'],
        super.fromJson(json);

  @override
  Map<String, dynamic> toJson() {
    Map<String, dynamic> superJson = super.toJson();
    superJson['post_id'] = post;
    return superJson;
  }

  @override
  String toString() {
    return '${user} - $createdAt - Post: ${post}';
  }
}

class ActivityRecordPostComment extends PostComment {
  final String? post;

  ActivityRecordPostComment({
    String? id,
    UserAbbr? user,
    String? content,
    String? createdAt,
    required this.post,
  }) : super(
      id: id,
      user: user,
      content: content,
      createdAt: createdAt
  );

  ActivityRecordPostComment.fromJson(Map<String, dynamic> json)
      : post = json['post'],
        super.fromJson(json);

  @override
  Map<String, dynamic> toJson() {
    Map<String, dynamic> superJson = super.toJson();
    superJson['post_id'] = post;
    return superJson;
  }

  @override
  String toString() {
    return '${user} - $createdAt - Activity Record: ${post}';
  }
}

class CreatePostComment {
  final String? userId;
  final String? postId;
  final String? content;

  CreatePostComment({
    required this.userId,
    required this.postId,
    required this.content,
  });

  Map<String, dynamic> toJson() {
    return {
      'user_id': userId,
      'post_id': postId,
      'content': content,
    };
  }

  @override
  String toString() {
    return "${toJson()}";
  }
}